text/BartekJachowicz/opis.tex

```diff
@@ -0,0 +1,129 @@
+Druga część to przepisanie ostatecznych wartości $L_{m-x-j+1}(j)$ które zapisane są w tablicy $R_1(j)$ do zwracanej tablicy $LL(j)$.\\
+
+
+Podobnie jak przy procedurze \textit{fillone} złożoność pamięciowa oraz czasowa są dosyć proste do uzasadnienia.
+Wykorzystanie pamięci to ponownie jedynie linowego rozmiaru tablice $R_1, r_2, LL$, zatem złożoność pamięciowa to $O(n + m)$.\\
+Złożoność czasowa to natomiast $(x + 1)$ wywołanie procedury \textit{fillone} w pętli for. Przypominając że złożoność procedury \textit{fillone} wynosi $O(n)$, zatem złożoność czasowa procedury \textit{calmid} to $O((x+1)n)$.
+
+\subsection{Znajdowanie perfekcyjnego cięcia}
+
+Procedura znajdowania \textit{perfekcyjnego cięcia} polega na dwukrotnym wywołaniu procedury calmid odpowiednio dla odwróconych słów $A, B$ oraz dla nieodwróconych. Po zakończeniu tych wywołań otrzymujemy w tablicach $LL_1, LL_2$ wyliczone wartości odpowiednio: $n + 1 - L_{m-w+1-k}(k)$ dla $k \leq r_1$ oraz $L_{w+k+1}(m-(w+k)-w')$\ $(= L_{m-w'+1-(p-k)}(p-k))$ dla $p-k \leq r_2$. Z twierdzenia 1.1 wynika zatem, że wystarczy zatem dobrać $u = k + w$ oraz $v = LL_1(k)$ by dostać poprawne perfekcyjne cięcie.\\
+Złożoność czasowa i pamięciowa to ta sama, która miała procedura calmid, gdyż dodatkowo wykonywana jest jedynie pętla \textit{for} by dobrać odpowiednie $(u, v)$, która ma złożoność $O(r_1) \leq O(n)$.
+
+\subsection{Złożoność czasowa i pamięciowa algorytmu}
+Patrząc na główną procedurę \textit{Longest common subsequence}, która działa zgodnie z techniką \textit{dziel i zwyciężaj} opisaną na początku dokumentu możemy obliczyć złożoność czasową algorytmu.
+Mamy do rozpatrzenia dwa przypadki:
+\begin{itemize}
+    \item Przypadku bazowego: jego rozwiązanie to wywołanie funkcji \textit{calmid} z parametrem $x = m-p$, gdzie $m - p \leq 2$, zatem jest to złożoność $O(n)$ oraz dodatkowe operacje w pętlach \textit{while}, których maksymalnie zostanie wykonanych $O(m)$. Zatem całkowita złożoność przypadku bazowego to $O(n + m)$.
+    \item Znajdowanie \textit{perfekcyjnego cięcia} oraz dwa wywołania rekurencyjne: pierwsze to dwa wywołania procedury \textit{calmid} z parametrem $x = w$ lub $x = w'$ odpowiednio, zatem ich złożoność to $O(n(w + 1))$ i $O(n(w' + 1))$. Złożoność ta to inaczej $O(n(m - p))$, ponieważ tak dobrany był parametry $w$ i $w'$. Dodając dwa wywołania rekurencyjne zgodne z \textit{perfekcyjnym cięciem} $(u, v)$ dostajemy złożoność:
+    \[
+        T(m, n, p) = O(n(m - p)) + T(u, v, u - w) + T(m - u, n - v, m - u - w')
+    \]
+    która po przeliczeniach okazuje się wynosić $O(n(m - p))$.
```

Comment:
At the first glance this works only if m = O(n). Please add that it has to be the case when $m - p \le 2$, since $p \le n$ (simple, but I was for a few moments puzzled).

Otherwise the base case would not fall into the $O(n(m - p))$ complexity, so the induction using the recurrence equation would not work.

Reply:
Done